src/connection.rs

```diff
@@ -477,29 +473,19 @@ impl<T: AsyncRead + AsyncWrite + Unpin> Connection<T> {
                 return Err(ConnectionError::Closed);
             }
 
-            let next_item = future::poll_fn(move |cx: &mut Context| {
-                let a = next_stream_command.poll_unpin(cx);
-                let b = next_control_command.poll_unpin(cx);
-                let c = next_frame.poll_unpin(cx);
-                if a.is_pending() && b.is_pending() && c.is_pending() {
-                    return Poll::Pending;
-                }
-                Poll::Ready((a, b, c))
-            });
-
-            let (stream_command, control_command, frame) = next_item.await;
-
-            if let Poll::Ready(cmd) = control_command {
-                self.on_control_command(cmd).await?
-            }
-
-            if let Poll::Ready(cmd) = stream_command {
-                self.on_stream_command(cmd).await?
-            }
-
-            if let Poll::Ready(frame) = frame {
-                if let Some(stream) = self.on_frame(frame.transpose().map_err(Into::into)).await? {
-                    return Ok(Some(stream));
+            let combined_future = future::select(
+                future::select(next_stream_command, next_control_command),
+                next_frame,
+            );
+            match combined_future.await {
+                Either::Left((Either::Left((cmd, _)), _)) => self.on_stream_command(cmd).await?,
+                Either::Left((Either::Right((cmd, _)), _)) => self.on_control_command(cmd).await?,
+                Either::Right((frame, _)) => {
+                    if let Some(stream) =
+                        self.on_frame(frame.transpose().map_err(Into::into)).await?
+                    {
+                        return Ok(Some(stream));
+                    }
                 }
```

Comment:
Have you considered using the `futures::select!` macro instead of nesting `future::select` calls?

Reply:
I have, though I favor `futures::future::select` over `futures::select!` due to the following reasons:

- While more verbose, I find `futures::future::select` easier to reason about than `futures::select!`.
- In general I prefer functions over macros.
- I understand the inner workings of `futures::future::select`. [The implementation](https://docs.rs/futures-util/0.3.21/src/futures_util/future/select.rs.html#101-113) is very simple.
- I don't fully understand the guarantees of `futures::select!`, e.g.:
    > If multiple futures are ready, one will be pseudo-randomly selected at runtime.
    
    https://docs.rs/futures/latest/futures/macro.select.html
    
    What happens to the result of the non-selected ready futures?